### PR TITLE
Vary bleed amount based on population density

### PR DESCRIPTION
**Pull request by quis**
There are basically two kinds of 4G masts:

Frequency | Range       | Bandwidth
----------|-------------|----------------------------------
800MHz    | Long (5km) | Low (can handle a bit of traffic)
1800Mhz   | Short (500m) | High (can handle lots of traffic)

The 1800Mhz masts are better in terms of how much traffic they can handle and how fast a connection they provide. But because they have quite short range, it’s only economical to install them in very built up areas†.

In more rural areas the 800MHz masts are better because they cover a wider area, and have enough bandwidth for the lower population density. 

The net effect of this is that cell broadcasts in rural areas are likely to bleed further, because the masts they are being broadcast from cover a wider area.

We can use population density as a proxy for how likely it is to be covered by 1800Mhz masts only, and therefore how much bleed we should expect. So this commit varies the amount of bleed shown based on the population density.

I came up with a formula to calculate this based on 3 ‘known’ data points:
- The most remote areas (for example the Scottish Highlands) should have the highest average bleed, estimated at 5km
- An town, like Crewe, should have about the same bleed as we were estimating before (1.5km) – Pete D thinks this is about right based on his knowledge of the area around his office in Crewe
- The most built up areas, like London boroughs, could have as little as 500m of bleed

This is the formula which roughly gives the right bleed distance (`b`) for each of their population densities (`d`):
```
b = 5900 - (log10(d) × 1_250)
```

[Plotted](https://docs.google.com/spreadsheets/d/13JPaWnAySaM8lnfXp7T3KjH77Fb---mwRDMoTTFGvgU/edit) on a curve it looks like this:

image

This is based on averages – remember that the UI shows where is _likely_ to receive the alert, based on bleed, not where it’s _possible_ to receive the alert. Here’s what it looks like on the map:

Before | After
---|---
Bleed area is a uniform 1,500m around the whole shape | Bleed area is wider in more rural areas and narrower in more built-up areas
Screenshot 2021-03-16 at 08 46 55 | Screenshot 2021-03-16 at 08 48 05
image |  image
image | image
Screenshot 2021-03-16 at 09 03 15 | Screenshot 2021-03-16 at 09 03 45

# Depends on

- [x] https://github.com/alphagov/notifications-utils/pull/848

---

† There are some additional subtleties which make this not strictly true:
- The 800Mhz masts are also used in built up areas to fill in the gaps between the areas covered by the 1800Mhz masts
- Switching between masts is inefficient, so if you’re moving fast through a built up area (for example on a train) your phone will only use the 800MHz masts so that you have to handoff from one mast to another less often

### CODE REVIEW

**Review by benthorner — COMMENTED** [review]
Phew, lots of reading and learning here!

The PR description is really helpful (the range stats at the top can be the other way round, by the way). Using phone density as a proxy for population density is a neat idea 🔢 .

I've got one more major comment about the edge case behaviour of the new bleed estimator.

File: tests/app/broadcast_areas/test_broadcast_area.py [hunk 1]
```diff
@@ -10,6 +12,10 @@
 )
 
 
+def close_enough(a, b):
```
[the remark this answers is not in the document]
Comment (quis) [reply tests/app/broadcast_areas/test_broadcast_area.py:15]: Can’t import from `tests/` in utils unfortunately because that directory is outside the package.

File: app/broadcast_areas/__init__.py [hunk 2]
```diff
@@ -68,6 +74,20 @@ def count_of_phones(self):
...
 
+    @property
+    def phone_density(self):
+        return self.count_of_phones / self.polygons.estimated_area
+
+    @property
+    def estimated_bleed_in_m(self):
+        if self.id.endswith(CITY_OF_LONDON.WARDS):
+            return 500
+        return 5_900 - (math.log(self.phone_density, 10) * 1_250)
```
[the remark this answers is not in the document]
Comment (quis) [reply app/broadcast_areas/__init__.py]: Limits are better than the London-specific thing, agreed. I‘ve changed it in 4367908

[the remark this answers is not in the document]
Comment (quis) [reply app/broadcast_areas/__init__.py]: Added a case for areas with `0` count of phones. We’ll return 1,500m, which is our best guess when we don’t know anything about the geography. See a74db6e

File: app/broadcast_areas/__init__.py [hunk 3]
```diff
@@ -68,6 +74,20 @@ def count_of_phones(self):
         # https://www.pivotaltracker.com/story/show/174837293
         return self._count_of_phones or 0
 
+    @property
+    def phone_density(self):
+        return self.count_of_phones / self.polygons.estimated_area
```
[the remark this answers is not in the document]
Comment (quis) [reply app/broadcast_areas/__init__.py:79]: Added a case for areas with `0` count of phones. We’ll return 1,500m, which is our best guess when we don’t know anything about the geography. See a74db6e

File: tests/app/broadcast_areas/test_broadcast_area.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply tests/app/broadcast_areas/test_broadcast_area.py:15]: Ah yeah, never mind. Thanks for checking.

File: app/broadcast_areas/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (benthorner) [reply app/broadcast_areas/__init__.py]: Thanks. 

(And the commit message has the more detailed thinking behind the curve 👍.)

**Review by benthorner — APPROVED** [review]
Looks good, thanks for addressing all my comments 💯.
